jplag.frontend-utils/src/main/java/de/jplag/SharedTokenType.java

```diff
@@ -0,0 +1,26 @@
+package de.jplag;
+
+/**
+ * Shared token types that occur for any language.
+ */
+public enum SharedTokenType implements TokenType {
```

Comment:
If you have something like `Boolean isFileEnd()` you can do the check on `Token::getType::isFileEnd` directly. 
However, the question is rather why we need that? What is the drawback of having token types that explicitly encapsulate this meaning? Since each type provides its own description, we don't need to include the shared token types in each language frontend anymore. Also, regarding the optionality of the separator token, this is automatically supported by the dynamic value assignment in the GST algorithm. If the separator token is absent in all submissions, no value is reserved for it.

Reply:
I would argue it makes for a cleaner architecture, as all token types for a language are defined in a single place. It is also easier for external frontend developers, as the interface reminds them to use EOF tokens in their parser implementation.
I personally find it cleaner not to have the shared token enum, but that might just be my view,

Reply:
From coding perspective, this will introduce a lot of duplicated code, as we will have the need to introduce e.g. an enum case in every language and implement the according method for indicating that that case is the file-end token type. Also, I would say that the language developer still defines all **language-related** tokens individually but those **metatypes** are shared and can be re-used. Thus I am not in favor of integrating the file-end token type into every language.
What we might consider: right now we do not use the `SEPARATOR` type at all. We could remove this and add some method to TokenType like `Boolean isExcludedFromMatching()` which defaults to false. Regarding the file-end token we can either enforce it by checking that the language frontend returns a sequence of tokens ending with the type file-end. Another approach might be to adapt the language frontend to expect `Map<File, List<Token>` without the file-end token. In the GST we then concat those lists using the FILE_END token. By this, the language developer does not need to care about the file-end token type at all.

Reply:
As the proposal with the `Boolean isExcludedFromMatching()` seemed very reasonable to me and also simplifies the GST algorithm, I already implemented it. So as of now, there is only `SharedTokenType#FILE_END` left.